### PR TITLE
Allow preamble macros to introduce declarations with documented names

### PR DESCRIPTION
**Pull request by DougGregor**
Preamble macros introduce code at the beginning of a function body.
Allow them to introduce declarations as well, so long as the macro
declaration properly declares the names it introduces. This allows all
sorts of exciting macros that introduce (e.g.) new local variables
that one can use.

### CODE REVIEW

**Comment by DougGregor** [comment]
Only the last commit is new

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Review by ktoso — APPROVED** [review]
Looking great!

File: test/Macros/macro_expand_body.swift [hunk 1]
```diff
@@ -27,36 +27,53 @@ extension String: ConjureRemoteValue {
...
 
-@available(SwiftStdlib 5.1, *)
-@Remote
-func f(a: Int, b: String) async throws -> String
+@Logged
+func useLogger() {
+  let x = 1
+  logger.log("use it")
+  print(x)
+}
```
Comment (ktoso) [inline test/Macros/macro_expand_body.swift:71]: Very cool, looking forward to this shape of those macros 🥳

**Pull request merged**